ecosystem/typescript/sdk/src/transaction_builder/builder.ts

```diff
@@ -292,3 +295,98 @@ export class TransactionBuilderABI {
     throw new Error("Invalid ABI.");
   }
 }
+
+export type RemoteABIBuilderConfig = Partial<Omit<ABIBuilderConfig, "sender">> & {
+  sender: MaybeHexString | AccountAddress;
+};
+
+interface AptosClientInterface {
+  getAccountModules: (arg0: string) => Promise<Gen.MoveModuleBytecode[]>;
+  getAccount: (accountAddress: MaybeHexString) => Promise<Gen.AccountData>;
+  getChainId: () => Promise<number>;
+}
+
+/**
+ * This transaction builder downloads JSON ABIs from the fullnodes.
+ * It then translates the JSON ABIs to the format that is accepted by TransactionBuilderABI
+ */
+export class TransactionBuilderRemoteABI {
+  // We don't want the builder to depend on the actual AptosClient. There might be circular dependencies.
+  constructor(
+    private readonly aptosClient: AptosClientInterface,
+    private readonly builderConfig: RemoteABIBuilderConfig,
+  ) {}
+
+  // Cache for 10 minutes
+  @MemoizeExpiring(10 * 60 * 1000)
+  async fetchABI(addr: string) {
+    const modules = await this.aptosClient.getAccountModules(HexString.ensure(addr).hex());
+    const abis = modules
+      .map((module) => module.abi)
+      .flatMap((abi) =>
+        abi.exposed_functions
+          .filter((ef) => ef.is_entry)
+          .map(
+            (ef) =>
+              ({
+                fullName: `${abi.address}::${abi.name}::${ef.name}`,
+                ...ef,
+              } as Gen.MoveFunction & { fullName: string }),
+          ),
+      );
+
+    const abiMap = new Map<string, Gen.MoveFunction & { fullName: string }>();
+    abis.forEach((abi) => {
+      abiMap.set(abi.fullName, abi);
+    });
+
+    return abiMap;
+  }
+
+  // Only support entry function payloads
+  async build(func: string, ty_tags: string[], args: any[]): Promise<RawTransaction> {
+    const [addr, module] = func.split("::");
+
+    // Downloads the JSON abi
+    const abiMap = await this.fetchABI(addr);
+    if (!abiMap.has(func)) {
+      throw new Error(`${func} doesn't exist.`);
+    }
+
+    const funcAbi = abiMap.get(func);
+
+    // Remove all `signer` and `&signer` from arugment list because the Move VM injects those arugments. Clients do not
+    // need to care about those args. `signer` and `&signer` are required be in the front of the arugment list. But we
+    // just loop through all arguments and filter out `singer` and `&signer`.
+    const originalArgs = funcAbi.params.filter((param) => param !== "signer" && param !== "&signer");
+
+    // Convert string arugments to TypeArgumentABI
+    const typeArgABIs = originalArgs.map((arg, i) => new ArgumentABI(`var${i}`, new TypeTagParser(arg).parseTypeTag()));
+
+    const scriptFunctionABI = new ScriptFunctionABI(
+      funcAbi.name,
+      ModuleId.fromStr(`${addr}::${module}`),
+      "",
+      funcAbi.generic_type_params.map((_, i) => new TypeArgumentABI(`type_tag_${i}`)),
```

Comment:
Having to do this:
```
TypeArgumentABI(`type_tag_${i}`)
```
Feels sorta janky, can we just make it take in `i`?

Reply:
TypeArgument in this case is useless. Whatever name is fine. just `i` should work too.